### PR TITLE
[SPARK-29513][SQL] REFRESH TABLE should look up catalog/table like v2 commands

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

Add RefreshTableStatement and make REFRESH TABLE go through the same catalog/table resolution framework of v2 commands.

### Why are the changes needed?

It's important to make all the commands have the same table resolution behavior, to avoid confusing end-users. e.g.

```
USE my_catalog
DESC t // success and describe the table t from my_catalog
REFRESH TABLE t // report table not found as there is no table t in the session catalog
```
### Does this PR introduce any user-facing change?

yes. When running REFRESH TABLE, Spark fails the command if the current catalog is set to a v2 catalog, or the table name specified a v2 catalog.

### How was this patch tested?

New unit tests

### CODE REVIEW

**Comment by imback82** [comment]
cc: @cloud-fan @rdblue @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala [hunk 1]
```diff
@@ -282,6 +282,10 @@ class ResolveSessionCatalog(
       AlterTableRecoverPartitionsCommand(
         v1TableName.asTableIdentifier,
         "MSCK REPAIR TABLE")
+
+    case RefreshTableStatement(tableName) =>
+      val v1TableName = parseV1Table(tableName, "REFRESH TABLE")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: We have `TableCatalog.invalidateTable` to implement a v2 REFRESH TABLE command.

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: Followed by `TableCatalog.loadTable`? Or`TableCatalog.loadTable` is called lazily (you don't call it inside `RefreshTableExec` - in this case, same behavior as `UNCACHE TABLE`)?

Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: Not followed by `loadTable`. It should invalidate and load the table lazily.

**Comment by rdblue** [comment]
Looks fine to me. We can add the implementation here or in a separate PR.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -1086,6 +1086,14 @@ class DataSourceV2SQLSuite
     }
   }
 
+  test("REFRESH TABLE: v2 table") {
+    val t = "testcat.ns1.ns2.tbl"
+    withTable(t) {
+      sql(s"CREATE TABLE $t (id bigint, data string) USING foo")
+      sql(s"REFRESH TABLE $t")
```
Comment (imback82) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: Any suggestion for a reasonable test here?

`ALTER TABLE` doesn't require `REFRESH TABLE` in the following existing test, correct?
```Scala
      sql(s"CREATE TABLE $t (id int, data string) USING $v2Format")
      sql(s"ALTER TABLE $t DROP COLUMN data")
      val table = getTableMetadata(t) // <- this calls TableCatalog.loadTable
```

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala (hunk 1 above)
Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala]: Thanks. I adde v2 command in this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: I think a reasonable test is to create a custom v2 implementation which implements `invalidateTable`.

**Review by cloud-fan — APPROVED** [review]

**Comment by imback82** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: We can implement `InMemoryTable. invalidateTable` and simply set a flag on a global object to indicate it has been called.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/RefreshTableExec.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.connector.catalog.{Identifier, TableCatalog}
+case class RefreshTableExec(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/RefreshTableExec.scala]: a blank line between import and case class.

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/RefreshTableExec.scala]: thanks. fixed.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala (hunk 2 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala]: Thanks for the suggestion! Exactly what I was planning to do. :)

**Review by viirya — APPROVED** [review]
looks good.

**Comment by cloud-fan** [comment]
LGTM, but we need to resolve conflicts again...

**Comment by imback82** [comment]
Just resolved conflicts. Thanks!

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by imback82** [comment]
Thanks all!

**Pull request closed without merging**